contracts/Strategy.sol

```diff
@@ -53,80 +222,624 @@ contract Strategy is BaseStrategy {
+
+        uint256 debt = vault.strategies(address(this)).totalDebt;
+
+        if (assets >= debt) {
+            _profit = assets.sub(debt);
+        } else {
+            _loss = debt.sub(assets);
+        }
+
+        _debtPayment = _debtOutstanding;
+        uint256 amountToFree = _debtPayment.add(_profit);
+
+        if (amountToFree > 0 && wantBal < amountToFree) {
+            (uint256 newLoose, ) = liquidatePosition(amountToFree.sub(wantBal));
+
+            // if we didnt free enough money, prioritize paying down debt before taking profit
+            if (newLoose < amountToFree) {
+                if (newLoose <= _debtPayment) {
+                    _profit = 0;
+                    _loss += _debtPayment.sub(newLoose);
+                    _debtPayment = newLoose;
+                } else {
+                    _profit = newLoose.sub(_debtPayment);
+                }
+            }
+        }
     }
 
     function adjustPosition(uint256 _debtOutstanding) internal override {
-        // TODO: Do something to invest excess `want` tokens (from the Vault) into your positions
-        // NOTE: Try to adjust positions so that `_debtOutstanding` can be freed up on *next* harvest (not immediately)
+        if (emergencyExit) {
+            return;
+        }
+
+        uint256 wantBalance = balanceOfWant();
+
+        uint256 shares = 0;
+
+        if (wantBalance > dustThreshold) {
+            (, shares) = depositInBento(wantBalance);
+        }
+
+        uint256 sharesInBento = sharesInBento();
+
+        if (sharesInBento > wantToBentoShares(dustThreshold, false)) {
+            // Get highest interest rate pair
+            uint256 highestInterestIndex =
+                highestInterestPairIndex(sharesInBento);
+
+            depositInKashiPair(highestInterestIndex, sharesInBento);
+        }
     }
 
     function liquidatePosition(uint256 _amountNeeded)
         internal
         override
         returns (uint256 _liquidatedAmount, uint256 _loss)
     {
-        // TODO: Do stuff here to free up to `_amountNeeded` from all positions back into `want`
-        // NOTE: Maintain invariant `want.balanceOf(this) >= _liquidatedAmount`
-        // NOTE: Maintain invariant `_liquidatedAmount + _loss <= _amountNeeded`
+        uint256 wantBalance = balanceOfWant();
+        if (_amountNeeded > wantBalance) {
+            uint256 amountToFree = _amountNeeded.sub(wantBalance);
+
+            uint256 deposited = estimatedTotalAssets().sub(wantBalance);
+
+            if (deposited < amountToFree) {
+                amountToFree = deposited;
+            }
+            if (amountToFree > 0) {
+                uint256 sharesNeeded = wantToBentoShares(amountToFree, true);
+                uint256 bentoShares = sharesInBento();
+
+                uint256 sharesToFreeFromKashi =
+                    bentoShares <= sharesNeeded
+                        ? sharesNeeded.sub(bentoShares)
+                        : 0;
```

Comment:
not sure im missing something, shouldn't this zero be all bentoshares instead if `sharesNeeded` is higher than bentoShares?

Reply:
If bentoshares are greater than sharesNeeded, we don't need to free anything from the kashiPair. Raw bento shares and shares in kashiPairs are accounted for separately.